components/match2/commons/bracket_template.lua

```diff
@@ -0,0 +1,146 @@
+		})
+	})
+end
+
+function BracketTemplate.readBracket(argsList)
+	local bracketDatas = Array.map(argsList, BracketTemplate.readBracketData)
+
+	local bracket = {
+		bracketDatasById = Table.map(
+			bracketDatas,
+			function(_, bracketData) return bracketData.matchId, bracketData end
+		),
+		rootMatchIds = {},
+		templateId = mw.title.getCurrentTitle().text,
+	}
+
+	-- Populate in bracketData.upperMatchId and rootMatchIds
+	local upperMatchIds = MatchGroupCoordinates.computeUpperMatchIds(bracket.bracketDatasById)
+	for _, bracketData in ipairs(bracketDatas) do
+		local upperMatchId = upperMatchIds[bracketData.matchId]
+		if not upperMatchId
+			and not String.endsWith(bracketData.matchId, 'RxMBR') then
+			table.insert(bracket.rootMatchIds, bracketData.matchId)
+		end
+		bracketData.upperMatchId = upperMatchId
+	end
+
+	-- Populate bracketData.coordinates
+	local bracketCoordinates = MatchGroupCoordinates.computeCoordinates(bracket)
+	for matchId, bracketData in pairs(bracket.bracketDatasById) do
+		bracketData.coordinates = bracketCoordinates.coordinatesByMatchId[matchId]
+	end
+
+	return bracket
+end
+
+function BracketTemplate.readBracketData(args)
+	local pageName = mw.title.getCurrentTitle().text
+	local function joinPageName(baseMatchId)
+		return pageName .. '_' .. baseMatchId
+	end
+
+	args.type = 'bracket'
+	local bracketData = MatchGroupUtil.bracketDataFromRecord(args)
+	bracketData.bracketResetMatchId = bracketData.bracketResetMatchId and joinPageName(bracketData.bracketResetMatchId)
+	bracketData.lowerMatchIds = Array.map(bracketData.lowerMatchIds, joinPageName)
+	bracketData.matchId = joinPageName(args.matchid)
+	bracketData.thirdPlaceMatchId = bracketData.thirdPlaceMatchId and joinPageName(bracketData.thirdPlaceMatchId)
+	bracketData.upperMatchId = bracketData.upperMatchId and joinPageName(bracketData.upperMatchId)
+	return bracketData
+end
+
+function BracketTemplate.store(bracket)
+	BracketTemplate.storeBracket(bracket)
+	BracketTemplate.storeDatapoint(bracket.templateId)
+end
+
+--[[
+Store bracket with placeholder match and opponent data in commons wiki LPDB
+]]
+function BracketTemplate.storeBracket(bracket)
+	for matchId, bracketData in pairs(bracket.bracketDatasById) do
+		local bracketId, baseMatchId = MatchGroupUtil.splitMatchId(matchId)
+		bracketData.matchId = nil
+
+		local opponent = {
+			name = MatchGroupUtil.matchIdToKey(baseMatchId),
+			type = 'literal',
+		}
+		local match = {
+			bracketid = bracketId,
+			match2bracketdata = MatchGroupUtil.bracketDataToRecord(bracketData),
+			matchid = baseMatchId,
+			opponent1 = opponent,
+		}
+		Match.store(match)
+	end
+end
+
+function BracketTemplate.storeDatapoint(templateId)
+	mw.ext.LiquipediaDB.lpdb_datapoint('ExtensionBracket_' .. templateId, {
+		type = 'extension bracket',
+	})
+end
+
+BracketTemplate.perfConfig = {
+	locations = {
+		'Module:Bracket/Template|*',
+		'Module:MatchGroup/Coordinates|*',
+		'Module:MatchGroup/Display/Bracket|*',
+	}
+}
```

Comment:
Used where?

Reply:
used for performance instrumentation

https://liquipedia.net/commons/Module:Performance/Util